### PR TITLE
feat(core): allow using hooks for interface entities

### PR DESCRIPTION
**Pull request by techmunk**

### CODE REVIEW

**Comment by techmunk** [comment]
I'm not sure the typing of hooks is 100% correct, but I'm not savvy enough with typescript to get it more accurate.

**Comment by techmunk** [comment]
Thanks @B4nan. I had considered the subscribers, however I was building this functionality onto a base entity and was hoping to keep it contextualized to that entity. Does the `getSubscribedEntities` method of an entity subscriber work for base entities?

**Comment by B4nan** [comment]
With a bit of metadata magic you could automatically list entity types that extends specific base entity. But the approach this PR allows seems to be much cleaner.

**Comment by B4nan** [comment]
Speaking of base entities, it would be probably good idea to add a test that defines a hook on base entity this way too.

**Comment by B4nan** [comment]
And we also need a mention about this new feature to the docs. I would add it to the entity schema docs, as that is the main use case for it.

**Comment by techmunk** [comment]
@B4nan I've pushed an update with some docs written and some tests on a base entity. I did have one strange issue with the base entity that I could not get to the bottom of though....

When the base entity schema was just an object (and not an EntitySchema instance), the hook functions would be in the hooks array, but would be `null`. Changing the schema to use `new EntitySchema()` instead, fixed this issue.

I don't really understand why, but I assume there's something in the entity discovery that works  a bit differently in those two cases.

**Comment by techmunk** [comment]
To be clear, only the functions would be null, the string method names were there and fine.

File: docs/docs/entity-schema.md [hunk 1]
```diff
@@ -183,3 +183,54 @@ export const schema = new EntitySchema<BookTag>({
...
+  }
+}
+
+// Defined outside of entity class - this bound to entity instance
+function beforeUpdate() {
+  this.version++;
+}
+
+// Defined outsite, this available via args.
+const beforeUpdate2 = (args: EventArgs) => args.entity.version++
```
Comment (B4nan) [inline docs/docs/entity-schema.md:221]:
```suggestion
const beforeUpdate2 = (args: EventArgs) => args.entity.version++;
```

File: docs/docs/entity-schema.md [hunk 2]
```diff
@@ -183,3 +183,54 @@ export const schema = new EntitySchema<BookTag>({
...
+
+export const schema = new EntitySchema<BookTag>({
```
Comment (B4nan) [inline docs/docs/entity-schema.md:223]: this is inferred from `class`

```suggestion
export const schema = new EntitySchema({
```

File: packages/core/src/events/EventManager.ts [hunk 3]
```diff
@@ -27,12 +27,12 @@ export class EventManager {
   dispatchEvent<T extends AnyEntity<T>>(event: EventType.onInit, args: Partial<EventArgs<T>>): unknown;
   dispatchEvent<T extends AnyEntity<T>>(event: EventType, args: Partial<EventArgs<T> | FlushEventArgs>): Promise<unknown>;
   dispatchEvent<T extends AnyEntity<T>>(event: EventType, args: Partial<EventArgs<T> | FlushEventArgs | TransactionEventArgs>): Promise<unknown> | unknown {
-    const listeners: [EventType, EventSubscriber<T>][] = [];
+    const listeners: (readonly [PropertyKey | NonNullable<EventSubscriber<T>[keyof EventSubscriber<T>]>, EventSubscriber<T> | AnyEntity<T>])[] = [];
```
Comment (B4nan) [inline packages/core/src/events/EventManager.ts:30]: please extract this into a new type, its too complex and reduces readabilty, same for the `runListener` usage

File: packages/core/src/events/EventManager.ts [hunk 4]
```diff
@@ -42,11 +42,18 @@ export class EventManager {
       }
     }
 
+    const runListener = ([hook, subscriber]: readonly [PropertyKey | NonNullable<EventSubscriber<T>[keyof EventSubscriber<T>]>, EventSubscriber<T> | AnyEntity<T>]) => {
+      if (typeof hook !== 'function') {
+        return subscriber[hook]!(args as (EventArgs<T> & FlushEventArgs & TransactionEventArgs));
```
Comment (B4nan) [inline packages/core/src/events/EventManager.ts:47]: let's extract `EventArgs<T> & FlushEventArgs & TransactionEventArgs` to a type too

**Review by B4nan — CHANGES_REQUESTED** [review]

**Comment by techmunk** [comment]
@B4nan Thanks. I was actually just sitting down to work on this and saw that you merged it! Cheers for all the hard work you put in!

**Comment by B4nan** [comment]
FYI just finished with simplifying the implementation as well as types a bit: https://github.com/mikro-orm/mikro-orm/commit/f3599292515941d3bef365160b18685eb15227d3

**Comment by techmunk** [comment]
That's a much cleaner implementation. :+1:

**Pull request merged**